### PR TITLE
Enabled authentication in golang based agents

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
PR contains the following changes that are separated in different commits:
* Adds an ability to enable auth for go agents with env variable
* Moves internal environment resolving to RuntimeInfrastructure
* Adds provisioning internal environment with configured installers
* Adds provisioning of common env vars to abstract infrastructure
* Adds provisioning of CHE_AUTH_ENABLED env var
* Adds usage of machine token while connecting to exec, terminal, workspace agents via websocket

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/6577

#### Release Notes
Enabled authentication in golang based agents

### CODE REVIEW

**Review by azatsarynnyy — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

**Review by akorneta — APPROVED** [review]

File: assembly-multiuser/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MultiUserCheWsMasterModule.java [hunk 1]
```diff
@@ -79,6 +79,8 @@ protected void configure() {
     bind(PreferenceDao.class).to(JpaPreferenceDao.class);
     bind(PermissionChecker.class).to(PermissionCheckerImpl.class);
 
+    bindConstant().annotatedWith(Names.named("che.agents.auth_enabled")).to(true);
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply assembly-multiuser/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MultiUserCheWsMasterModule.java:82]: Do you have an idea how to name this property?

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/jsonrpc/ExecAgentJsonRpcInitializer.java [hunk 2]
```diff
@@ -104,7 +110,8 @@ private void internalInitialize(String machineName) {
               execAgentServer.ifPresent(
                   server ->
                       initializer.initialize(
-                          machine.getName(), singletonMap("url", server.getUrl())));
+                          machine.getName(),
+                          singletonMap("url", wsAgentURLModifier.modify(server.getUrl()))));
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/jsonrpc/ExecAgentJsonRpcInitializer.java]: I'd just fixed code to work in right way =) I'll try to refactor it

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/CheWsAgentURLModifier.java [hunk 3]
```diff
@@ -12,18 +12,18 @@
...
+import org.eclipse.che.ide.api.workspace.model.RuntimeImpl;
 
 /**
  * Dummy implementation.
  *
  * @author Anton Korneta
  */
 @Singleton
-public class CheWsAgentLinksModifier implements WsAgentURLModifier {
+public class CheWsAgentURLModifier implements WsAgentURLModifier {
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/CheWsAgentURLModifier.java]: It was implemented in this way before my changes. Makes sense to remove interface here

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java [hunk 4]
```diff
@@ -98,6 +109,22 @@ protected void configure() {
     bind(org.eclipse.che.api.workspace.server.WorkspaceService.class);
     install(new FactoryModuleBuilder().build(ServersCheckerFactory.class));
 
+    Multibinder<InternalEnvironmentProvisioner> internalEnvironmentProvisioners =
+        Multibinder.newSetBinder(binder(), InternalEnvironmentProvisioner.class);
+    internalEnvironmentProvisioners.addBinding().to(InstallerConfigProvisioner.class);
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java:114]: Can you explain what do you mean?

File: agents/go-agents/terminal-agent/main.go [hunk 5]
```diff
@@ -148,11 +149,18 @@ func (cfg *terminalAgentConfig) registerFlags() {
 	)
 
 	// auth configuration
+	defaultAuthEnabled := false
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply agents/go-agents/terminal-agent/main.go:152]: No =) I'll try to add them

File: agents/go-agents/exec-agent/main.go [hunk 6]
```diff
@@ -175,11 +176,18 @@ func (cfg *execAgentConfig) registerFlags() {
 	)
 
 	// auth configuration
+	defaultAuthEnabled := false
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply agents/go-agents/exec-agent/main.go:179]: Makes sense. I'll take a look and rework it

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/terminal/TerminalPresenter.java [hunk 7]
```diff
@@ -107,7 +111,7 @@ public void connect() {
                                     "Machine "
                                         + machine.getName()
                                         + " doesn't provide terminal server."));
-                connectToTerminal(terminalServer.getUrl());
+                connectToTerminal(wsAgentURLModifier.modify(terminalServer.getUrl()));
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/terminal/TerminalPresenter.java]: I think renaming it to AgentURLModifier would be better

File: assembly-multiuser/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MultiUserCheWsMasterModule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (garagatyi) [reply assembly-multiuser/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MultiUserCheWsMasterModule.java:82]: I don't have a suggestion, so let's go with the current variant.

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java:114]: As far as I understand you're talking about the case when agents are not installed by the installers but provided by the container(machine). So, this case is not covered in che6 branch, I wouldn't rework it in the scope of this PR.

[the remark this answers is not in the document]
Comment (garagatyi) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java:114]: I mean that if a server is provided by the image, not installer this configuration would influence it.

File: agents/go-agents/exec-agent/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply agents/go-agents/exec-agent/main.go:179]: The method is renamed to `init`. Is it OK for you?

File: agents/go-agents/terminal-agent/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply agents/go-agents/terminal-agent/main.go:152]: go files modified by me are not covered with tests at all. I'd like to cover them but I need time to learn more about testing in go. 
So, I'd leave it without tests as there were no tests before and I add a really small piece of code.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/CheWsAgentURLModifier.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/CheWsAgentURLModifier.java]: `WsAgentURLModifier` interface replaces with `AgentURLModifier` class. Please take a look

File: agents/go-agents/exec-agent/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (garagatyi) [reply agents/go-agents/exec-agent/main.go:179]: :) ok

File: agents/go-agents/exec-agent/main.go [hunk 8]
```diff
@@ -25,14 +25,15 @@ import (
...
+	"strconv"
 )
 
 var (
 	config = &execAgentConfig{}
 )
 
 func init() {
-	config.registerFlags()
+	config.init()
```
Comment (garagatyi) [inline agents/go-agents/exec-agent/main.go:36]: Does it make sense to have this method?

File: agents/go-agents/terminal-agent/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (garagatyi) [reply agents/go-agents/terminal-agent/main.go:152]: ok

File: agents/go-agents/exec-agent/main.go (hunk 8 above)
Comment (sleshchenko) [reply agents/go-agents/exec-agent/main.go:36]: I think yes

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/RuntimeInfrastructure.java [hunk 9]
```diff
@@ -143,4 +155,93 @@ protected abstract void internalEstimate(InternalEnvironment env)
...
+          throw new InfrastructureException(
+              String.format(
+                  "Installer '%s' contains server '%s' conflicting with machine configuration",
+                  installer.getId(), serverEntry.getKey()));
+        }
+      }
+    }
+  }
+
+  //TODO Take a look | Add tests
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/RuntimeInfrastructure.java]: Are you going to add tests in this PR?

Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/RuntimeInfrastructure.java]: It is already fixed by https://github.com/eclipse/che/pull/6983/commits/b39856235c7b33d839e9f6258d433e6721eb3c9f

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/provision/env/EnvVarEnvironmentProvisioner.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.Set;
+import javax.inject.Inject;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalEnvironment;
+import org.eclipse.che.api.workspace.server.spi.provision.InternalEnvironmentProvisioner;
+import org.eclipse.che.commons.lang.Pair;
+
+/**
+ * @author Alexander Garagatyi
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/provision/env/EnvVarEnvironmentProvisioner.java]: Can you add javadocs here?

**Review by garagatyi — APPROVED** [review]
Looks good. Nice refactoring!

**Comment by sleshchenko** [comment]
@vparfonov @azatsarynnyy I've update commit with changes that are needed to using machine token by IDE. Please review https://github.com/eclipse/che/pull/6983/commits/ec0786d36ed9f407392ba6f2eb2b754071d19cec

**Review by vparfonov — APPROVED** [review]

**Review by azatsarynnyy — APPROVED** [review]

**Pull request merged**